Monads/Option.cs

```diff
@@ -133,4 +133,138 @@ public class Some<T>(T value) : Option<T>(value) where T : class;
     /// </summary>
     /// <typeparam name="T">the type of the value that is missing</typeparam>
     public class None<T>() : Option<T>() where T : class;
+
+    /// <summary>
+    /// The monad type used to represent that there might be a value, but doesn't have to be as well.
+    /// </summary>
+    /// <remarks>Also provides simple shorthands for ways of manipulating with them.</remarks>
+    /// <typeparam name="T">the underlying value</typeparam>
+    public class ValueOption<T> : IEquatable<ValueOption<T>>
+        where T : struct
+    {
+        protected T? _value;
+
+        /// <summary>
+        /// Returns the underlying value
+        /// </summary>
+        /// <returns>the contained value</returns>
+        /// <exception cref="InvalidOperationException">there is no underlying value</exception>
+        public T Value => _value ?? throw new InvalidOperationException();
+
+        /// <summary>
+        /// Returns a value indicating whether the underlying value or is <see langword="null"/> or not.
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if the underlying value exists.
+        /// <see langword="false"/> if there is no underlying value.
+        /// </returns>
+        //[System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
+        [System.Diagnostics.CodeAnalysis.MemberNotNullWhen(true, nameof(_value))]
+        public bool HasValue => _value is not null;
+
+        protected ValueOption(T? value = null) => _value = value;
+
+        /// <summary>
+        /// Creates an <see cref="ValueOption{T}"/> that contains a value of type <typeparamref name="T"/>
+        /// </summary>
+        /// <param name="value">value to store</param>
+        /// <returns>An <see cref="ValueOption{T}"/> with some value of type <typeparamref name="T"/></returns>
+        public static ValueOption<T> Some(T value) => new(value);
+
+        /// <summary>
+        /// Creates an <see cref="ValueOption{T}"/> with no underlying value
+        /// </summary>
+        /// <returns>An <see cref="ValueOption{T}"/> containing nothing</returns>
+        public static ValueOption<T> None() => new(null);
+
+        /// <summary>
+        /// Tries to commit an operation with the possible underlying value.
+        /// </summary>
+        /// <typeparam name="R">Type of the result of the operation</typeparam>
+        /// <param name="func">operation to perform</param>
+        /// <returns>
+        /// If there was an underlying value, returns an <see cref="ValueOption{R}.Some(R)"/>.
+        /// If there was no value, returns an <see cref="ValueOption{R}.None"/>.
+        /// </returns>
+        public ValueOption<R> Bind<R>(Func<T, R> func) where R : struct
+            => HasValue ? new(func(_value!.Value)) : new();
```

Comment:
No bang suffix required, compiler should know from the `HasValue` property that the member shall not be `null`
```suggestion
            => HasValue ? new(func(_value.Value)) : new();
```